src/lib.rs

```diff
@@ -0,0 +1,534 @@
+    /// This argument is only meaninful is memlock is required
+    pub allow_mem_resize: bool,
+    /// Whether mulithreading is enabled
+    pub allow_multithread: bool,
+    /// Whether Memtester returns immediately if a test fails or continues until all tests are run
+    pub allow_early_termination: bool,
+}
+
+#[derive(Debug)]
+pub struct MemtestReportList {
+    pub tested_usize_count: usize,
+    pub mlocked: bool,
+    pub reports: Vec<MemtestReport>,
+}
+
+#[derive(Debug)]
+pub struct MemtestReport {
+    pub test_type: MemtestType,
+    pub outcome: Result<MemtestOutcome, MemtestError>,
+}
+
+/// An struct to ensure the test timeouts in a given duration
+#[derive(Clone, Debug)]
+struct TimeoutChecker {
+    deadline: Instant,
+    curr_test: Option<TimeoutCheckerTestInstance>,
+}
+
+// TODO: A better name for this struct
+/// A struct for storing states of TimeoutChecker specific to a test instance
+#[derive(Clone, Debug)]
+struct TimeoutCheckerTestInstance {
+    test_start_time: Instant,
+    expected_iter: u64,
+
+    completed_iter: u64,
+    checkpoint: u64,
+    num_checks_completed: u128,
+    checking_interval: Duration,
+}
+
+impl Memtester {
+    // TODO: Memtester without given base_ptr, ie. take care of memory allocation as well
+
+    /// Create a Memtester containing all test types in random order
+    pub fn all_tests_random_order(args: MemtesterArgs) -> Memtester {
+        let mut test_types = vec![
+            MemtestType::TestOwnAddress,
+            MemtestType::TestRandomVal,
+            MemtestType::TestXor,
+            MemtestType::TestSub,
+            MemtestType::TestMul,
+            MemtestType::TestDiv,
+            MemtestType::TestOr,
+            MemtestType::TestAnd,
+            MemtestType::TestSeqInc,
+            MemtestType::TestSolidBits,
+            MemtestType::TestCheckerboard,
+            MemtestType::TestBlockSeq,
+        ];
+        test_types.shuffle(&mut thread_rng());
+
+        Self::from_test_types(args, test_types)
+    }
+
+    /// Create a Memtester with specified test types
+    pub fn from_test_types(args: MemtesterArgs, test_types: Vec<MemtestType>) -> Memtester {
+        Memtester {
+            test_types,
+            timeout: args.timeout,
+            require_memlock: args.require_memlock,
+            allow_working_set_resize: args.allow_working_set_resize,
+            allow_mem_resize: args.allow_mem_resize,
+            allow_multithread: args.allow_multithread,
+            allow_early_termination: args.allow_early_termination,
+        }
+    }
+
+    /// Run the tests, possibly after locking the memory
+    pub fn run(&self, memory: &mut [usize]) -> anyhow::Result<MemtestReportList> {
+        // TODO: Should have a minimum memory length so that we don't UB when `memory.len()` is too small
+        let mut timeout_checker = TimeoutChecker::new(Instant::now(), self.timeout);
```

Comment:
Hmm... Yeah. Perhaps it makes more sense to just calculate `deadline` here and pass it around to each thread/test, and every test gets it's own newly-created `TimeoutChecker`.

The point of us getting the `TimeoutChecker` out of the individual tests was for 2 reasons: 1) So you don't have to repeat yourself creating it for each test, and 2) More importantly -- It allows the caller to inject their own `TimeoutChecker` implementations into your test functions.